### PR TITLE
front: Pathfinding module

### LINKED ISSUE #2699: front: pathfinding: Status module proposal to concentrate all about pathfinding & UX guidelines
The idea is to have a module dedicated to pathfinding's status and more !
:warning:  the module MUST BE AS SMALL AS POSSIBLE :warning: why not above map module ?
It presents all existing pathfinding state & infos about the path founded.

_Don't bother too much with UI details if you want, I will later put colors & glitter if necessary ;)_

### 1. Pathfinding is running

All modules implicated in pathfinding configuration must be disabled. My proposal is to use the same system as used in new modal CRUD Infra : a conditionnal `<div />` in absolute position with a white background and half opacity with a loader above. It forbids changing pathfinding parameters and avoid to launch another pathfinding before the first is complete.

Pathfinding's module disabled should only display the message that a pathfinding is running and a cancel button that cancel the request.

### 2. Pathfinding has an error

The error message is displayed, with reasons if possible.

### 3. Pathfinding is done

No need to display a message, just display pathfinding ID (for debug purpose, will be removed on prod), and distance of the path in kilometers.

### 4. Pathfinding is impossible
This is not an error, only an information indicating why pathfinding can't be launched.
- missing rolling stock
- missing origin
- missing destination

### 5. :tada: NEW :tada: Pathfinding should be launched
For now, pathfinding is launched automatically when rs/origin/destination are complete or changed
This is the case too when a waypoint is added : this is a problem, because adding 5 waypoints creates 5 new pathfindings.
We should stop launching a path when a waypoint is added/modified/removed, instead, the frontend knows that configuration has changed, and pathfinding module displays an enough visible alert that pathfinding has to be relaunched, with a dedicated button.

### PR DESCRIPTION
**Pull request by anisometropie**
closes #2699

### CODE REVIEW

File: front/public/locales/fr/operationalStudies/manageTrainSchedule.json [hunk 1]
```diff
@@ -28,6 +28,16 @@
     "unableToRetrieveTimetable": "Récupération grille horaire impossible",
     "unableToRetrieveTimetableList": "Récupération liste de grilles horaires impossible"
   },
+  "pathfinding": "Statut recherche de chemin",
```
Comment (nicolaswurtz) [inline front/public/locales/fr/operationalStudies/manageTrainSchedule.json]: be more specific: pathFinding should mean "Recherche d'itinéraire", if you want "statut", add "status"

File: front/public/locales/fr/operationalStudies/manageTrainSchedule.json [hunk 2]
```diff
@@ -28,6 +28,16 @@
...
+  "pathfindingInProgress": "Recherche du chemin en cours",
+  "pathfindingError": "Erreur dans la recherche du chemin : ",
+  "pathfindingDone": "Recherche terminée : ",
+  "pathfindingMissingParams": "Éléments manquants pour la recherche : ",
```
Comment (nicolaswurtz) [inline front/public/locales/fr/operationalStudies/manageTrainSchedule.json]: Only if relevant, use `{{ variable }}` formatting possibilities https://www.i18next.com/translation-function/formatting

File: front/public/locales/fr/operationalStudies/manageTrainSchedule.json [hunk 3]
```diff
@@ -28,6 +28,16 @@
...
+  "No path could be found": "Aucun chemin trouvé",
+  "No path could be found after loading Gauge constraints": "Aucun chemin trouvé avec ce gabarit de matériel",
+  "No path could be found after loading Electrification constraints": "Aucun chemin trouvé un matériel avec ce type d’électrification",
```
Comment (nicolaswurtz) [inline front/public/locales/fr/operationalStudies/manageTrainSchedule.json]: Translation a bit chelou :smile:

File: front/src/applications/operationalStudies/components/ManageTrainSchedule/Itinerary/DisplayItinerary/Origin.tsx [hunk 4]
```diff
@@ -49,31 +46,18 @@ function Origin(props: OriginProps) {
...
         <RiMapPin2Fill />
       </span>
       <span>{t('origin')}</span>
-      <button
-        type="button"
-        onClick={openModalWrapperBecauseTypescriptSucks}
-        className="btn btn-link"
-      >
-        <small className="ml-1">{pathfindingID}</small>
-      </button>
```
Comment (nicolaswurtz) [inline front/src/applications/operationalStudies/components/ManageTrainSchedule/Itinerary/DisplayItinerary/Origin.tsx:76]: Okay it was not the better name :D But the functionnality is needed, it gives the possibility to export json pathFinding data for research purpose, I will explain to you on Slack. Please always ask before completely remove a functionality, thx :)

File: front/src/common/Pathfinding/pathfinding.scss [hunk 5]
```diff
@@ -0,0 +1,70 @@
+.pathfinding-main-container {
```
Comment (nicolaswurtz) [inline front/src/common/Pathfinding/pathfinding.scss]: why not put scss file in /styles/common ? And create a common.scss file who add each file inside /styles/common/ ?

File: front/src/common/Pathfinding/pathfinding.scss [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  .pathfinding-error {
+    @extend .pathfinding-subsection;
+    color: red;
+  }
+
+  .missing-params {
+    @extend .pathfinding-subsection;
+    color: #ffb612;
```
Comment (nicolaswurtz) [inline front/src/common/Pathfinding/pathfinding.scss]: please use Bootstrap's color with var(--orange) for example

File: front/src/common/Pathfinding/pathfinding.scss [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    .details {
+      font-weight: 400;
+    }
+  }
+
+  .pathfinding-error {
+    @extend .pathfinding-subsection;
+    color: red;
```
Comment (nicolaswurtz) [inline front/src/common/Pathfinding/pathfinding.scss]: please use Bootstrap's color with var(--red) for example

File: front/src/common/Pathfinding/pathfinding.scss [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    display: flex;
+    flex-direction: row;
+    align-items: center;
+  }
+
+  .pathfinding-done {
+    @extend .pathfinding-subsection;
+    
+    .title {
+      color: green;
```
Comment (nicolaswurtz) [inline front/src/common/Pathfinding/pathfinding.scss]: please use Bootstrap's color with var(--green) for example

File: front/src/common/Pathfinding/pathfinding.scss [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    display: flex;
+    flex-direction: column;
+    justify-content: center;
+    align-items: center;
+
+    .pathfinding-in-progress-card {
+      width: 50%;
+      border-radius: 4px;
+      padding: 1rem 1.5rem;
+      background: white;
```
Comment (nicolaswurtz) [inline front/src/common/Pathfinding/pathfinding.scss]: please use Bootstrap's color with var(--white) for example

**Review by nicolaswurtz — CHANGES_REQUESTED** [review]
Very important : keep modal access with JSON from pathfinding

File: front/src/common/Pathfinding/pathfinding.scss (hunk 8 above)
Comment (anisometropie) [reply front/src/common/Pathfinding/pathfinding.scss]: done

File: front/src/common/Pathfinding/pathfinding.scss (hunk 9 above)
Comment (anisometropie) [reply front/src/common/Pathfinding/pathfinding.scss]: done

File: front/src/common/Pathfinding/pathfinding.scss (hunk 6 above)
Comment (anisometropie) [reply front/src/common/Pathfinding/pathfinding.scss]: done

File: front/src/common/Pathfinding/pathfinding.scss (hunk 5 above)
Comment (anisometropie) [reply front/src/common/Pathfinding/pathfinding.scss]: why not :)

File: front/src/styles/scss/_variables.scss [hunk 10]
```diff
@@ -12,4 +12,9 @@
   --coolgray5: #b9b9b9;
   --coolgray3: #d7d7d7;
   --coolgray1: #f2f2f2;
+  --warning-orange: #ffb612; // from text-warning bootstrap class
+  --error-red: red;
+  --success-green: green;
+  --modal-gray-transparent-background: rgba(0,0,0,0.1);
+  --card-background-white: white;
```
Comment (nicolaswurtz) [inline front/src/styles/scss/_variables.scss]: I'm sorry, we must have misunderstood each other, no need to create new variables they already exist!

Please use them :)

image

File: front/src/common/Pathfinding/Pathfinding.tsx [hunk 11]
```diff
@@ -38,7 +38,7 @@ import {
 
 import ModalPathJSONDetail from 'applications/operationalStudies/components/ManageTrainSchedule/Itinerary/ModalPathJSONDetail';
 
-import './pathfinding.scss';
+import 'styles/scss/common/pathfinding.scss';
```
Comment (nicolaswurtz) [inline front/src/common/Pathfinding/Pathfinding.tsx]: import it & put it inside `src/styles/...`

File: front/src/styles/scss/common/pathfinding.scss [hunk 12]
```diff
@@ -11,22 +11,23 @@
     @extend .pathfinding-subsection;
     
     .title {
-      color: green;
+      color: var(--success-green);
```
Comment (nicolaswurtz) [inline front/src/styles/scss/common/pathfinding.scss]: same as above

File: front/src/styles/scss/common/pathfinding.scss [hunk 13]
```diff
@@ -11,22 +11,23 @@
...
     .details {
+      margin-left: 0.25rem;
       font-weight: 400;
     }
   }
 
   .pathfinding-error {
     @extend .pathfinding-subsection;
-    color: red;
+    color: var(--error-red);
```
Comment (nicolaswurtz) [inline front/src/styles/scss/common/pathfinding.scss]: same as above

File: front/src/styles/scss/common/pathfinding.scss [hunk 14]
```diff
@@ -11,22 +11,23 @@
...
   }
 
   .missing-params {
     @extend .pathfinding-subsection;
-    color: #ffb612;
+    color: var(--warning-orange);
```
Comment (nicolaswurtz) [inline front/src/styles/scss/common/pathfinding.scss]: same as above

File: front/src/styles/scss/common/pathfinding.scss [hunk 15]
```diff
@@ -36,7 +37,7 @@
     left: 0;
     min-width: 100%;
     min-height: 100%;
-    background-color: rgba(0,0,0,0.1);
+    background-color: var(--modal-gray-transparent-background);
```
Comment (nicolaswurtz) [inline front/src/styles/scss/common/pathfinding.scss]: same as above

File: front/src/styles/scss/common/pathfinding.scss [hunk 16]
```diff
@@ -48,7 +49,7 @@
       width: 50%;
       border-radius: 4px;
       padding: 1rem 1.5rem;
-      background: white;
+      background: var(--card-background-white);
```
Comment (nicolaswurtz) [inline front/src/styles/scss/common/pathfinding.scss]: same as above

**Review by nicolaswurtz — CHANGES_REQUESTED** [review]

**Review by alexandredamiron — COMMENTED** [review]
in ./.storybook/preview.js, can you update the import paths ? It cause storybook build crash and will make the CI fail.

**Review by nicolaswurtz — DISMISSED** [review]

**Review by flomonster — DISMISSED** [review]

**Review by nicolaswurtz — APPROVED** [review]

**Pull request merged**